### PR TITLE
Inference test improvements

### PR DESCRIPTION
**Pull request by ashao**
The co-located inference tests have been improved:
- a Fortran version of the inference test has been implemented and tests
- The C++ client was changed to use an array of random numbers for inference
- An MPI_Barrier was put at start of each iteration of the test to ensure that all ranks are synchronized at the beginning and make each iteration comparable to any other
- 'warm up' the database by doing a round of inference before the main loop

### CODE REVIEW

File: fortran-inference/inference_scaling_imagenet.F90 [hunk 1]
```diff
@@ -0,0 +1,252 @@
+program main
```
Comment (amandarichardsonn) [inline fortran-inference/inference_scaling_imagenet.F90:1]: I might have to trust you on this file lol

File: fortran-inference/inference_scaling_imagenet.F90 [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+character(len=255) :: timing_file
+
+! I/O Variables
+integer :: timing_unit
+
+! SmartRedis variables
+type(client_type) :: client
+character(len=255) :: script_key, model_key
+
+! Retrieve all environment variables
```
Comment (amandarichardsonn) [inline fortran-inference/inference_scaling_imagenet.F90:31]: Is "SR_LOG_FILE": "srlog.out" and "SR_LOG_LEVEL": "info" needed here?

File: imagenet/model_saver.py [hunk 3]
```diff
@@ -10,7 +10,9 @@ def save_model(device: str = "CPU"):
             raise Exception("Requested to create GPU model, but CUDA was not found.")
     # Device's torch name
     device_name = "cuda" if device.startswith("GPU") else "cpu"
-    model = models.resnet50(pretrained=True)
+    # model = models.resnet50(pretrained=True)
```
Comment (amandarichardsonn) [inline imagenet/model_saver.py:13]: take these comments put?

Comment (amandarichardsonn) [reply imagenet/model_saver.py:13]: out*

File: utils.py [hunk 4]
```diff
@@ -165,7 +166,8 @@ def setup_resnet(model, device, num_devices, batch_size, address, cluster=True):
                                        model,
                                        "TORCH",
                                        device,
-                                       batch_size)
+                                       batch_size,
+                                       min_batch_size=batch_size)
```
Comment (amandarichardsonn) [inline utils.py:170]: why is there a batch_size var and a min batch size var, with the batch size assigned to min batch size? can this var be taken away in the smartsim api?

File: fortran-inference/inference_scaling_imagenet.F90 (hunk 1 above)
Comment (amandarichardsonn) [reply fortran-inference/inference_scaling_imagenet.F90:1]: comment file?

File: figures/plot_colocated_inference.ipynb
Comment (amandarichardsonn) [inline figures/plot_colocated_inference.ipynb:1]: remove the hash-able dict func in this file if not being used

File: cpp-inference/inference_scaling_imagenet.cpp [hunk 5]
```diff
@@ -141,12 +140,37 @@ void run_mnist(const std::string& model_name,
     script_key = "resnet_script_0";
   }
 
-  MPI_Barrier(MPI_COMM_WORLD);
   if (!rank)
     std::cout<<"All ranks have Resnet image"<<std::endl;
 
+  // Warmup the database with an inference
```
Comment (amandarichardsonn) [inline cpp-inference/inference_scaling_imagenet.cpp:146]: why is it necessary to warm up the db?

File: driverinference/main.py [hunk 6]
```diff
@@ -275,31 +283,33 @@ def _create_inference_session(cls,
                         num_devices=num_devices)
 
         return model, resnet_model
-    
+
     @classmethod
     def _create_colocated_inference_session(cls,
```
Comment (amandarichardsonn) [inline driverinference/main.py:288]: in this file -> add language to the infer sess name, so users dont have to look at the csv file to check if fortran or cpp

File: fortran-inference/inference_scaling_imagenet.F90 (hunk 1 above)
Comment (amandarichardsonn) [inline fortran-inference/inference_scaling_imagenet.F90:1]: The output times in the fortran test have spacing that look like `2,put_tensor,   0.6258940696716309E-01` -> the cpp csv lines look like `0,unpack_tensor,0.0587351`. Im wondering if the space between the function name and time for the fortran csv code will affect the performance results parse.

Comment (amandarichardsonn) [reply fortran-inference/inference_scaling_imagenet.F90:1]: client() and main() do not have these spaces in fortran csv files

File: fortran-inference/inference_scaling_imagenet.F90 [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+! Timing variables
+real(kind=8) :: main_start, main_end
+real(kind=8) :: client_start, client_end
+real(kind=8) :: delta_t
+character(len=255) :: timing_file
+
+! I/O Variables
+integer :: timing_unit
+
+! SmartRedis variables
```
Comment (amandarichardsonn) [inline fortran-inference/inference_scaling_imagenet.F90:27]: maybe add some comments below like here

File: driverinference/main.py [hunk 8]
```diff
@@ -10,7 +10,7 @@
 
 
 class Inference:
- 
+
```
Comment (amandarichardsonn) [inline driverinference/main.py:13]: maybe add language to run arguments in output like below

Run Arguments:
        constraint = P100
        nodes = 1
        ntasks = 18
        ntasks-per-node = 18
        language = ur mom

Comment (amandarichardsonn) [reply driverinference/main.py:13]: I see it in the Executable path tho, but would be nice to have it right in my face

File: cpp-inference/inference_scaling_imagenet.cpp [hunk 9]
```diff
@@ -119,8 +119,7 @@ void run_mnist(const std::string& model_name,
   for(int i=0; i<224; i++)
```
Comment (amandarichardsonn) [inline cpp-inference/inference_scaling_imagenet.cpp:119]: what is the 244?

Comment (amandarichardsonn) [reply cpp-inference/inference_scaling_imagenet.cpp:119]: the char length? What happens if the char length is not fully needed, does that add extra runtime?

File: utils.py [hunk 10]
```diff
@@ -151,7 +151,8 @@ def setup_resnet(model, device, num_devices, batch_size, address, cluster=True):
                                             model,
                                             "TORCH",
                                             0, num_devices,
-                                            batch_size)
+                                            batch_size,
+                                            min_batch_size=batch_size)
         client.set_script_from_file_multigpu("resnet_script_0",
```
Comment (amandarichardsonn) [inline utils.py:156]: this doesnt need to be "resnet_script_{i}" like below does it?

Comment (amandarichardsonn) [reply utils.py:156]: hmmm well its not in a for loop lol so ignore this

File: imagenet/model_saver.py [hunk 11]
```diff
@@ -10,7 +10,9 @@ def save_model(device: str = "CPU"):
...
+    model = models.shufflenet_v2_x0_5(pretrained=True)
```
Comment (amandarichardsonn) [inline imagenet/model_saver.py:14]: this does not need to be models.resnet?

**Review by amandarichardsonn — APPROVED** [review]
Got it runnin and completed, really awesome work this is amazing really, I have a couple questions that I added below and I encourage more comments! But LGTM!

File: imagenet/model_saver.py (hunk 11 above)
Comment (amandarichardsonn) [reply imagenet/model_saver.py:14]: also remove the comments #model if not needed!

File: imagenet/model_saver.py (hunk 3 above)
Comment (amandarichardsonn) [reply imagenet/model_saver.py:13]: This comment is written like 2 other times I am so sorry

File: fortran-inference/inference_scaling_imagenet.F90 (hunk 2 above)
Comment (ashao) [reply fortran-inference/inference_scaling_imagenet.F90:31]: Good call

**Pull request merged**